### ISSUE TITLE
[WIP] Componentise the app.

### ISSUE THREAD

**Issue by jdickey** [jdickey/new_poc#54]
By which I mean,

> Package the various components of the application, such as DSOs, entities, and so on, into either standalone Gems or [unbuilt Rails dependencies](http://pivotallabs.com/unbuilt-rails-dependencies-how-to-design-for-loosely-coupled-highly-cohesive-components-within-a-rails-application/) which are a functional near-equivalent.

I'd argue that this is an *extremely* desirable activity, as it proves that we can more easily use and intermix the components developed here with other applications, whether public or internal. Specifically, to drive the point home, this project was originally started as a *research activity* supporting commercial product development. Being able to use the code for that purpose as reliably and conveniently as possible is obviously important.

**Comment by jdickey** [comment]
With the close of Issues #150 and #152, we're off to a good start on what should get easier and more iterative as we go along. Ryan McGeary was right, nearly four years ago; [*"Vendor Everything" still applies*](http://ryan.mcgeary.org/2011/02/09/vendor-everything-still-applies/).With Mike Barinek of Pivotal's riff on [unbuilt dependencies](http://pivotallabs.com/unbuilt-rails-dependencies-how-to-design-for-loosely-coupled-highly-cohesive-components-within-a-rails-application/), you can choose whether, and when, to package and/or release Gems to vendor at a time of your choosing. Win all the way round.

**Comment by jdickey** [comment]
Right, then; I think we've figured out a few things.

1. DAOs, initially implemented as Rails/ActiveRecord models inside our app, as well as the controllers that mediate access to them, are *fundamentally tied to Rails apps* (more specifically, engines, of which a Rails app is an instance. Attempting to package a model/DAO or controller as a standalone Gem is, seemingly *by design*, an exercise in futility;
1. That having been said, there is still obvious value in extracting our Repository and Entity layers. The app (and controller and DAO) would depend on (list in its Gemfile) the Gems for the Repository layer, which would themselves depend on the Entity-layer Gems. This makes sense once you realise that the Entity layer, containing the bulk of the domain logic, should not depend on the implementation layers;
1. That brings up controller-initiated Actions. While it *should* be possible (and desirable) to extract these as well, I can live with them being tied into the app (with the controllers that use them).

This should probably drive a rewording of the original issue definition, if not the README for the project.

**Comment by jdickey** [comment]
Now that we've aborted #154, let's start at the *other* end of the chain it was proposing to walk. See Issue #160.

**Comment by jdickey** [comment]
So… Commit 51d7ee0 merged PR #163 into `master`, closing the retroactively-limited #160 and opening #166 in the process. Which way to go next, doing a new #160-like issue for `UserEntity`, or backtracking and fixing #166?

I'm choosing the former, and here's why: I *believe* I understand how to retrofit the text-post body-builder class. Writing that same sort of code (and ideally extracting it into one or more modules that can be shared among all concerned) will *prove* it. It can also easily be argued that migrating `UserEntity` is a prerequisite to closing *this* issue, whereas cleaning up the now-existing support code is tech debt but not blocking.

*Having said that*, closing this issue without having first closed #166 would be a grievous misjudgement.

**Comment by jdickey** [comment]
OK, so #171 and #172 have been suspended, labelled `wontfix` and effectively abandoned for the foreseeable future, at least by [me](https://github.com/jdickey/). DAOs, *as* ActiveRecord models, must remain part of the delivery-system megalith barring the use of heavy equipment (engines) or deep magic to which I am not presently adept (getting ActiveRecord happy in a theoretically-standalone Gem).

It should be possible for you, the intrepid explorer with time to burn, to investigate [FriendlyId](https://github.com/norman/friendly_id) and similar ActiveRecord-dependent Gems and figure out how they accomplish this. I've other components to pull out (like Repositories) that I'm confident won't require that level of effort. Low-hanging fruit as compared to moon shots, and all that.

**Comment by jdickey** [comment]
...and there's no point extracting [`PostDao`](https://github.com/jdickey/new_poc/blob/3767361/app/daos/post_dao.rb) and [`UserDao`](https://github.com/jdickey/new_poc/blob/3767361/app/daos/user_dao.rb) into unbuilt dependencies/standalone Gems:

* `PostRepository` is literally just a single three-line method overriding and calling [`Newpoc::Repository::Base#initialize`](https://github.com/jdickey/new_poc/blob/3767361/lib/newpoc-repository-base/lib/newpoc/repository/base.rb#L11-L13), specifying the (default, overridable) entity factory and repository classes to use ([`PostFactory`](https://github.com/jdickey/new_poc/blob/3767361/app/entity_factories/post_factory.rb) and [`PostDao`](https://github.com/jdickey/new_poc/blob/3767361/app/daos/post_dao.rb) respectively);
* `UserRepository` is only slightly more involved. It has all of two public API methods: [`#all`](https://github.com/jdickey/new_poc/blob/3767361/app/repositories/user_repository.rb#L12-L14), which is a one-line filter of the method it overrides; and [`#authenticate`](https://github.com/jdickey/new_poc/blob/3767361a91c9d87893aea9720e74a0158a618d35/app/repositories/user_repository.rb#L16-L25), which makes extensive, implementation-level use of the DAO. (Remember that we just failed to extract a DAO into an unbuilt Gem; see Issue #171.)

**tl;dr:** Repositories are fairly tightly coupled to DAOs, at least for Rails (due to its lack of encapsulation/isolation of internal components). As such, they "live" wherever DAOs live, which for now is the Rails app.

On to the next thing.

**Comment by jdickey** [comment]
The point of the above is that extracting *our own* bespoke architectural levels (such as Entities) into packaged components is just fine, as long as any such component depends on no Rails component more low-level than ActiveModel. Neither ActiveRecord nor `ActiveRecord::Base` subclass ActiveModel; they just pull in numerous ActiveModel modules.

It should be possible to extract most view-helper code, in that the current implementation of most helper global functions are thin wrappers around classes, which can themselves be extracted. But DAOs/models, controllers and (likely) view templates (which Rails terms as *views*) are apparently welded to the delivery-app internals.

**Comment by jdickey** [comment]
This whole endeavour is turning out to be a *spectacularly* bad idea. Pivotal's [unbuilt-Rails-dependency](http://pivotallabs.com/unbuilt-rails-dependencies-how-to-design-for-loosely-coupled-highly-cohesive-components-within-a-rails-application/) scheme is likely a Godsend for projects that

* can be readily split into [Rails](http://guides.rubyonrails.org/engines.html) [engines](http://railscasts.com/episodes/277-mountable-engines); and
* see no likelihood in those engines being released/maintained separately from their originating app.

Components that don't meet *both* those qualifications, either because they're more granular than engines (which can really be thought of as "embedded apps within your Rails app"), and/or because they are likely to be shared with other apps *or other Gems*, *should* be packaged *as Gems*. If publication as open source via [RubyGems.org](https://rubygems.org/) or similar public repositories is inappropriate for whatever reason (e.g., disclosure of proprietary information), private services like [Gemfury](https://gemfury.com/l/gem-server) offer a [variety of plans](https://gemfury.com/plans). Using standard build/package/release workflows for (fully-independent at this point) Gems brings several benefits:

* Shared code (boilerplate, etc.) can be DRYed up and isolated in support Gems that provide a "fire-and-forget" solution to repeated problems. Examples of this include Actions' [success/failure event-broadcasting conventions](https://github.com/jdickey/new_poc/blob/3d5cf1b/lib/newpoc-action-post-edit/lib/newpoc/action/post/edit.rb#L86-L92); processing steps which are identical in numerous Actions (such as [`#prohibit_guest_access`](https://github.com/jdickey/new_poc/blob/3d5cf1b/lib/newpoc-action-post-edit/lib/newpoc/action/post/edit.rb#L94-L97); and so on;
* Higher-level Gems can have direct dependencies on lower-level code and eliminate that management hassle from the enclosing app. *Each and every* "unbundled" Gem must now be included in the hosting app's `Gemfile`;
* Testing can be *dramatically* simplified through the use of existing, independently-verified code rather than extremely brittle, difficult mocks. The proverbial last straw that inspired *this* write-up was the [mock](https://github.com/jdickey/new_poc/blob/3d5cf1b/lib/newpoc-action-user-update/spec/newpoc/action/user/update_spec.rb#L44-L66) in the `Newpoc::Action::User::Update` specs that really, *really* wanted to be replaced by a thin shim around [`Newpoc::Repository::Base`](https://github.com/jdickey/new_poc/tree/3d5cf1b/lib/newpoc-repository-base), which does all the heavy lifting in the original.

----

We learned during our early dalliance with following along the path of [*Objects on Rails*](http://objectsonrails.com/) that mocking is a Spectacularly Bad Idea at scale; more effort goes into developing and testing your *mocks* (and keeping them up-to-date) than goes into the code that uses those mocks (or into the code that the mocks mock). We were able to keep the idea afloat long enough to get the first 12 of 15 Actions identified in [this comment](https://github.com/jdickey/new_poc/issues/173#issuecomment-71854385) on Issue #173 done, in some *15 man-days of effort*. The last three, however, are a slog *because of the mind games we have to play with the tests*.

----

What would be a better approach to isolating components, enforcing (and proving enforcement of) their isolation from other code in the application? I'm presently leaning towards the approach recommended by folks such as Nick Sutterer (in his Trailblazer [architecture](https://github.com/apotonick/trailblazer) and [book](https://leanpub.com/trailblazer)), that

* encloses components such as our Actions that are tied to traditional Rails components (controllers, in Actions' case) in namespaces within the containing class. For example, the (former) `Actions::UpdateUser` class (now [`Newpoc::Action::User::Update`](https://github.com/jdickey/new_poc/blob/3d5cf1b/lib/newpoc-action-user-update/lib/newpoc/action/user/update.rb) would become `UsersController::Actions::Update` or even `UsersController::UpdateAction`;
* Those enclosed components would expose their public API *only*; implementation of each public method would call into an `Internals` namespace within the component class (e.g., `UsersController::UpdateAction::Internals`, which would be locked away within the component class as a `private_constant`, preventing outside access (including from the containing controller).

Other components, not so intimately tied (such as what is now [`Newpoc::Repository::Base`](https://github.com/jdickey/new_poc/tree/3d5cf1b/lib/newpoc-repository-base) could be made into standalone, utility Gems. Proprietary logic, if any, could be injected idiomatically.

This gets our current house of cards into a more stable, durable structure before the wind picks up any more than it already has (and before I get any more burnt out than I already am, which is indeed possible).

**Comment by jdickey** [comment]
So, six weeks after really starting out on this mega-WIP, we're gone over the proverbial cliff, done a [classic Wile E Coyote manoeuvre](http://youtu.be/_d8ROhH3_vs), and come up with our *next* plan to solve this problem.

Read through the immediate preceding comment and you'll understand most of what you need to. First off is the apparently lowest-hanging fruit, extracting what was originally a [`RepositoryBase` class](https://github.com/jdickey/new_poc/blob/66f4211/app/repositories/repository_base.rb) before being transformed into a [`Newpoc::Repository::Base` class in an unbuilt Gem](https://github.com/jdickey/new_poc/tree/3d5cf1b/lib/newpoc-repository-base); it will now be explicitly published as a `repository_base` Gem. This will require minor tweaking of the latest code (changing naming conventions, mostly) along with *significant* changes to the existing unbuilt-Gem's `README`.

**Comment by jdickey** [comment]
Well, *that* (Issue #202 and PR #204) took entirely too long. Breaking out "oh, by the way, since we're in here we might as well get *this* done too" work would have made more visible artefacts of progress at the (probable) cost of longer overall time, but that's an argument for another day.

On to moving Actions *into* child namespaces of controllers, ditching the pseudo-Gems for code that isn't directly *shared* in the first place (though it would be easy enough to reverse that decision now that we have the craft technique down to a near-science). I'll be opening new issues/PRs (referencing this issue) for each.

**Comment by jdickey** [comment]
So, before we start ripping out the "Gems", let's tackle the actions we never got pulled out *into* unbuilt Gems:

- [ ] `CreateUser`;
- [ ] `CreatePost`.

(`UpdateUser` *was* converted but the obsolete code hasn't yet been removed; we'll do that action after the other two, cleaning as we go.)

**Comment by jdickey** [comment]
Continuing to rip out the unbundled-dependency pseudo-Gems for Actions, starting with User actions:

- [ ] `Newpoc::Action::User::Index`;
- [ ] `Newpoc::Action::User::New`;
- [ ] `Newpoc::Action::User::Edit`; and
- [ ] `Newpoc::Action::User::Show`.

**Comment by jdickey** [comment]
Actions for Posts:

- [ ] `Newpoc::Action::Post::Index';
- [ ] `Newpoc::Action::Post::New';
- [ ] `Newpoc::Action::Post::Show';
- [ ] `Newpoc::Action::Post::Edit';
- [ ] `Newpoc::Action::Post::Update'.

Once I get these moved, I really should look at factoring out common behaviour between Post and User actions. Several of them, perhaps most visibly `Index`, differ *minimally*.

**Comment by jdickey** [comment]
As I said,

>  I really should look at factoring out common behaviour between Post and User actions.

On further consideration, I'll defer *that* until I work on Issue #162, which is morphing into a general "clean up all the things" issue.

----

On to the `SessionsController` and its related actions; moving

- [ ] `Newpoc::Action::Session::Destroy` to `SessionsController::Action::Destroy`;
- [ ] `Newpoc::Action::Session::New` to `SessionsController::Action::New`;
- [ ] `Newpoc::Action::Session::Create` to `SessionsController::Action::Create`.

Once *those* are done, we'll be down to the entities and the Markdown-to-HTML converter service.

**Comment by jdickey** [comment]
So, what's left? There's the [`MarkdownHtmlConverter` service](https://github.com/jdickey/new_poc/tree/6e5bbe4/lib/newpoc-services-markdown_html_converter), the [`Post` entity](https://github.com/jdickey/new_poc/tree/6e5bbe4/lib/newpoc-entity-post) and the [`User` entity](https://github.com/jdickey/new_poc/tree/6e5bbe4/lib/newpoc-entity-user), all of which *should* be straightforward. (Bear in mind that we thought that about what became Issue #232 as well, though.)

So what are we really *accomplishing* with this whole junket, clawing back the code that so laboriously was put into unbuilt-dependency "Gems"? For one thing, it's a *lot* easier to share code while keeping things separated now. We see, for example, that the code in [`lib/action_support`](https://github.com/jdickey/new_poc/tree/6e5bbe4/lib/action_support) would make reasonable sense as a Gem, since one or more classes are used in *every* Action class. We've got clearer lines, both of separation and of dependence, between different bits of code than we had a couple of months ago. We've isolated some bits of code that are hard to test in the usual fashion, and either worked around the difficulties or called them out explicitly as something that *ought* to be addressed eventually. (Issue #232 again.)

What we *haven't* done, yet at least, is to address the separation of roles/use cases involved in the two main entity classes, `Post` and `User`. Both classes have accreted behaviours (validation, presentation logic) that could/ideally should be broken out per [ISP](https://drive.google.com/file/d/0BwhCYaYDn8EgOTViYjJhYzMtMzYxMC00MzFjLWJjMzYtOGJiMDc5N2JkYmJi/view) (and hence SRP). But it *should* be pretty clear how a real, non-POC implementation would take those ideas forward.

What's the best use of time at this point? Might we simply declare victory and get out, perhaps after renaming this issue as "Componentise *the actions within* the app", leaving the entities and remaining service as exercises for the future? That would allow us to continue on with the other [things worth fixing](https://github.com/jdickey/new_poc/issues?q=is%3Aissue+is%3Aopen+-label%3Afixlater+-label%3Awontfix).

We've long since met the [original goals](https://github.com/jdickey/new_poc/tree/6e5bbe4b92759199589c8d6dc96489061d3bfa0b#what-is-this-success-you-speak-of) of this POC. We've taken some good ideas from Avdi Grimm's [*Objects on Rails*](http://objectsonrails.com/) book; seen how dated a lot of its details are; embraced [various](https://github.com/apotonick/trailblazer) [other](http://dci-in-ruby.info/) [sources]https://github.com/krisleech/wisper) of [inspiration](http://commandercoriander.net/blog/2014/10/02/isolating-active-record/); and adapted the whole m&eacute;lange to give us an app that uses Rails, as opposed to a Rails app, that should be a *lot* easier to bring new developers with classic Ruby or OO experience up to speed on than we started out with. (Having an architecture that is less structurally monolithic and more adaptable as performance needs warrant is the *other* goal of the project that *should* have been met by now.)

Thoughts?

**Comment by jdickey** [comment]
Now that we've moved `Newpoc::Entity::User` to `Entity::User`, it's on to the corresponding transition for Post entities. This *will* take less time.
